File: FutureVoice/Views/VoiceComparisonSheet.swift
import SwiftUI

/// A/B the recording against the clone built from it.
///
/// Exists because "it doesn't sound like me" was, until now, a question the
/// app asked and gave the learner no way to answer. The meet act played the
/// clone once, then offered "record again?" — so the judgement was made
/// against a memory of their own voice, and a memory of your own voice is the
/// one recording everybody finds strange. Put the two side by side on the
/// SAME sentence and the question answers itself in ten seconds.
///
/// Both sides speak the opening of the script they actually read, so the
/// comparison is like for like. The clone's line is synthesized once and
/// cached in `PhraseAudioStore` under (text, voiceId), so reopening this sheet
/// — or opening it again after a re-record — costs nothing the second time.
struct VoiceComparisonSheet: View {
    @EnvironmentObject private var appState: AppState
    @Environment(\.dismiss) private var dismiss

    /// The opening of the script the learner read. Passed in rather than
    /// derived here: only the caller knows whether the take was read in the
    /// native or the target language.
    let scriptOpening: String
    /// Called when they decide the clone isn't them. The caller owns what
    /// happens next (the meet act walks back to the script step).
    var onRerecord: () -> Void

    /// How much of the recording to play. The clone's line is the script's
    /// opening ~120 characters, which lands around ten seconds of speech; the
    /// recording is the same words read at the learner's own pace, so a fixed
    /// window lines the two up closely enough to judge. It never has to be
    /// exact — what's being compared is a voice, not a performance.
    private static let sampleWindow: TimeInterval = 12

    private enum Side { case mine, clone }

    @StateObject private var minePlayer = AudioPlayer()
    @StateObject private var clonePlayer = AudioPlayer()

    @State private var cloneAudio: Data?
    @State private var loading = false
    @State private var error: String?

    var body: some View {
        NavigationStack {
            List {
                compareSection
                verdictSection
            }
            .navigationTitle("Same words, both voices")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Done") { stopAll(); dismiss() }
                }
            }
            .task { await loadCloneLine() }
            .onDisappear { stopAll() }
            .alert("Couldn't play the comparison", isPresented: Binding(
                get: { error != nil },
                set: { if !$0 { error = nil } }
            )) {
                Button("OK") { error = nil }
            } message: {
                Text(error ?? "")
            }
        }
    }

    // MARK: - The two voices

    private var compareSection: some View {
        Section {
            HStack(spacing: 12) {
                sideButton(.mine)
                sideButton(.clone)
            }
            .padding(.vertical, 4)
            Text("\u{201C}\(scriptOpening)\u{201D}")
                .font(.footnote)
                .foregroundStyle(.secondary)
        } footer: {
            VStack(alignment: .leading, spacing: 8) {
                Text(explain("Tap one, then the other. Listen for the voice, not the words — everyone finds their own recording strange at first."))
                // The comparison speaks whatever the recording holds, which
                // for a native-script take is NOT the language being learned.
                // That looks like a mismatch and is the whole point: most
                // "this isn't me" comes from hearing yourself in a language
                // you don't own yet, not from a bad clone. Matching here
                // separates the two, so say so rather than leaving the learner
                // to wonder why they're being played Korean.
                if let read = readLanguage, read != targetCode {
                    Text(explain("The recording is in \(LanguageCatalog.learnerName(read)), so the comparison is too. If the two match here, then what sounds unlike you in another language is the language, not the voice."))
                }
            }
        }
    }

    private var readLanguage: String? {
        appState.cloneScriptLanguage.flatMap { LanguageCatalog.language($0)?.code ?? $0 }
    }
    private var targetCode: String {
        LanguageCatalog.language(appState.targetLanguage)?.code ?? appState.targetLanguage
    }

    private func sideButton(_ side: Side) -> some View {
        let isPlaying = playing(side)
        let ready = side == .mine ? sampleData != nil : cloneAudio != nil
        return Button {
            toggle(side)
        } label: {
            VStack(spacing: 8) {
                if side == .clone && loading {
                    ProgressView()
                        .frame(height: 30)
                } else {
                    Image(systemName: isPlaying ? "stop.fill" : "play.fill")
                        .font(.title2)
                        .frame(height: 30)
                }
                Text(side == .mine ? chrome("My recording") : chrome("The made voice"))
                    .font(.footnote.weight(.medium))
                    .multilineTextAlignment(.center)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .frame(maxWidth: .infinity)
            .frame(minHeight: 76)
            .contentShape(Rectangle())
        }
        .buttonStyle(.bordered)
        .controlSize(.large)
        .tint(isPlaying ? Color.accentColor : nil)
        .disabled(!ready)
    }

    // MARK: - The decision

    private var verdictSection: some View {
        Section {
            Button {
                stopAll()
                dismiss()
            } label: {
                Label("It's me — keep it", systemImage: "checkmark")
            }
            Button(role: .destructive) {
                stopAll()
                // Dismiss first: the caller walks this screen back to the
                // script step, and a sheet still up over that transition
                // leaves the learner tapping through a stale comparison.
                dismiss()
                onRerecord()
            } label: {
                Label("Not me — record again", systemImage: "mic.fill")
            }
        } footer: {
            Text(explain("Recording again during setup is free, and the voice you have now is only replaced if you keep the new one."))
        }
    }

    // MARK: - Playback

    private var sampleData: Data? { VoiceSampleStore.shared.url.flatMap { try? Data(contentsOf: $0) } }

    private func playing(_ side: Side) -> Bool {
        side == .mine ? minePlayer.isPlaying : clonePlayer.isPlaying
    }

    /// One at a time, always — comparing means alternating, and two voices at
    /// once is just noise.
    private func toggle(_ side: Side) {
        if playing(side) { stopAll(); return }
        stopAll()
        do {
            switch side {
            case .mine:
                guard let data = sampleData else { return }
                minePlayer.prepare(data)
                // The take runs 60–90 s; only its opening is the same words
                // the clone is about to say.
                minePlayer.playSegment(from: 0, to: Self.sampleWindow, loop: false)
            case .clone:
                guard let data = cloneAudio else { return }
                try clonePlayer.play(data, source: "voice_comparison")
            }
        } catch {
            self.error = error.localizedDescription
        }
    }

    private func stopAll() {
        minePlayer.stop()
        clonePlayer.stop()
    }

    /// The clone saying the same opening. Cache first — this sheet is opened
    /// exactly when someone is unsure, which is also when they open it twice.
    private func loadCloneLine() async {
        guard cloneAudio == nil, !loading, let voiceId = appState.voiceCloneId else { return }
        if let cached = PhraseAudioStore.shared.data(text: scriptOpening, voiceId: voiceId) {
            cloneAudio = cached
            return
        }
        loading = true
        defer { loading = false }
        do {
            // The same model the meet act's greeting used. A likeness
            // judgement has to hear the clone at its best, and the result is
            // cached per line, so the fidelity model's higher per-character
            // cost is paid once.
            let data = try await ElevenLabsClient.shared.synthesize(
                voiceId: voiceId, text: scriptOpening,
                modelId: ElevenLabsClient.fidelityModelId, purpose: "voice_comparison")
            _ = PhraseAudioStore.shared.save(data, text: scriptOpening, voiceId: voiceId)
            cloneAudio = data
        } catch {
            self.error = error.localizedDescription
        }
    }
}

File: android/app/src/main/java/com/roro/futurevoice/ui/VoiceComparisonSheet.kt
package com.roro.futurevoice.ui

import android.media.AudioAttributes
import android.media.MediaPlayer
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Stop
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.roro.futurevoice.R
import com.roro.futurevoice.audio.Mp3Player
import com.roro.futurevoice.data.AuthRepository
import com.roro.futurevoice.net.ElevenLabsClient
import com.roro.futurevoice.talk.VoiceCloneScript
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

/**
 * A/B the recording against the clone built from it — iOS
 * `VoiceComparisonSheet`. "It doesn't sound like me" is judged against a
 * memory of one's own voice, which everybody finds strange; put the two side
 * by side on the SAME sentence and the question answers itself.
 *
 * Both sides speak the opening of the script the learner read. The clone's
 * line is synthesized once (fidelity model, cached on disk under the voice),
 * so reopening costs nothing.
 */
object VoiceComparison {
    fun sampleFile(filesDir: File) = File(filesDir, "voice/clone-sample.wav")
    fun exists(filesDir: File) = sampleFile(filesDir).length() > 0

    /** The script's first paragraph, trimmed to ~10 s of speech as on iOS. */
    fun opening(targetLanguage: String): String {
        val text = VoiceCloneScript.paragraphs(targetLanguage).firstOrNull() ?: ""
        if (text.length <= 120) return text
        val head = text.take(160)
        val stop = head.indexOfLast { it in ".?!。？！" }
        return if (stop > 40) head.substring(0, stop + 1) else head
    }

    /** How much of the recording to play: the take runs 60–90 s; only its
     *  opening is the same words the clone says. */
    const val SAMPLE_WINDOW_MS = 12_000L
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun VoiceComparisonSheet(voiceId: String, targetLanguage: String, onRerecord: () -> Unit, onDismiss: () -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val opening = remember(targetLanguage) { VoiceComparison.opening(targetLanguage) }
    val clonePlayer = remember { Mp3Player(context.cacheDir, source = "voice_comparison") }
    var mine by remember { mutableStateOf<MediaPlayer?>(null) }
    var mineJob by remember { mutableStateOf<Job?>(null) }
    var playingMine by remember { mutableStateOf(false) }
    var playingClone by remember { mutableStateOf(false) }
    var cloneAudio by remember { mutableStateOf<ByteArray?>(null) }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    fun stopAll() {
        mineJob?.cancel(); mineJob = null
        mine?.let { runCatching { it.stop() }; it.release() }; mine = null
        playingMine = false
        clonePlayer.stop(); playingClone = false
    }
    DisposableEffect(Unit) { onDispose { stopAll() } }

    // Cache first — this sheet is opened exactly when someone is unsure,
    // which is also when they open it twice.
    LaunchedEffect(voiceId) {
        val cache = File(context.filesDir, "voice/comparison-${voiceId.hashCode()}.mp3")
        if (cache.length() > 0) { cloneAudio = withContext(Dispatchers.IO) { cache.readBytes() }; return@LaunchedEffect }
        loading = true
        runCatching {
            ElevenLabsClient(AuthRepository()).synthesize(voiceId, opening,
                modelId = ElevenLabsClient.FIDELITY_MODEL_ID, purpose = "voice_comparison")
        }.onSuccess { data ->
            cloneAudio = data
            withContext(Dispatchers.IO) { cache.parentFile?.mkdirs(); cache.writeBytes(data) }
        }.onFailure { error = it.message ?: "" }
        loading = false
    }

    // One at a time, always — comparing means alternating.
    fun toggleMine() {
        if (playingMine) { stopAll(); return }
        stopAll()
        runCatching {
            val mp = MediaPlayer().apply {
                setAudioAttributes(AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH).build())
                setDataSource(VoiceComparison.sampleFile(context.filesDir).absolutePath)
                setOnCompletionListener { playingMine = false }
                prepare(); start()
            }
            mine = mp; playingMine = true
            mineJob = scope.launch { delay(VoiceComparison.SAMPLE_WINDOW_MS); if (playingMine) stopAll() }
        }.onFailure { error = it.message ?: "" }
    }
    fun toggleClone() {
        if (playingClone) { stopAll(); return }
        stopAll()
        val data = cloneAudio ?: return
        playingClone = true
        scope.launch { runCatching { clonePlayer.play(data) }.onFailure { error = it.message ?: "" }; playingClone = false }
    }

    error?.let {
        AlertDialog(onDismissRequest = { error = null }, title = { Text(stringResource(R.string.couldn_t_play_the_comparison)) },
            text = { Text(it) }, confirmButton = { TextButton(onClick = { error = null }) { Text(stringResource(R.string.ok)) } })
    }
    ModalBottomSheet(onDismissRequest = { stopAll(); onDismiss() }) {
        Column(Modifier.fillMaxWidth().navigationBarsPadding().padding(horizontal = 20.dp).padding(bottom = 28.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)) {
            Text(stringResource(R.string.same_words_both_voices), style = MaterialTheme.typography.titleLarge, fontWeight = FontWeight.SemiBold)
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                Side(stringResource(R.string.my_recording), playingMine, ready = VoiceComparison.exists(context.filesDir), loading = false,
                    Modifier.weight(1f)) { toggleMine() }
                Side(stringResource(R.string.the_made_voice), playingClone, ready = cloneAudio != null, loading = loading,
                    Modifier.weight(1f)) { toggleClone() }
            }
            Text("\u201C$opening\u201D", style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            GroupedFooter(stringResource(R.string.tap_one_then_the_other_listen_for_the_voice_not_the_words_ev_f4a3b5))
            GroupedSectionSpacer()
            GroupedCard {
                Row(Modifier.fillMaxWidth().padding(4.dp), verticalAlignment = Alignment.CenterVertically) {
                    TextButton(onClick = { stopAll(); onDismiss() }, modifier = Modifier.fillMaxWidth()) {
                        Icon(Icons.Filled.Check, contentDescription = null); Text("  " + stringResource(R.string.it_s_me_keep_it))
                    }
                }
                GroupedRowDivider(inset = false)
                Row(Modifier.fillMaxWidth().padding(4.dp), verticalAlignment = Alignment.CenterVertically) {
                    // Dismiss first: the caller walks to the record step, and a
                    // sheet still up over that transition is a stale comparison.
                    TextButton(onClick = { stopAll(); onDismiss(); onRerecord() }, modifier = Modifier.fillMaxWidth()) {
                        Icon(Icons.Filled.Mic, contentDescription = null, tint = MaterialTheme.colorScheme.error)
                        Text("  " + stringResource(R.string.not_me_record_again), color = MaterialTheme.colorScheme.error)
                    }
                }
            }
            GroupedFooter(stringResource(R.string.recording_again_during_setup_is_free_and_the_voice_you_have_75a508))
        }
    }
}

@Composable
private fun Side(label: String, playing: Boolean, ready: Boolean, loading: Boolean, modifier: Modifier, onClick: () -> Unit) {
    OutlinedButton(onClick = onClick, enabled = ready, modifier = modifier.heightIn(min = 76.dp)) {
        Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(6.dp)) {
            if (loading) CircularProgressIndicator(Modifier.size(22.dp), strokeWidth = 2.dp)
            else Icon(if (playing) Icons.Filled.Stop else Icons.Filled.PlayArrow, contentDescription = null)
            Text(label, style = MaterialTheme.typography.labelLarge)
        }
    }
}
